src/rgw/rgw_kms.cc

```diff
@@ -250,6 +251,114 @@ class KvSecretEngine: public VaultSecretEngine {
 
 };
 
+class KmipSecretEngine;
+class KmipGetTheKey {
+private:
+	CephContext *cct;
+	std::string work;
+	bool failed = false;
+	int ret;
+protected:
+	KmipGetTheKey(CephContext *cct) : cct(cct) {}
+	KmipGetTheKey& keyid_to_keyname(boost::string_view key_id);
+	KmipGetTheKey& get_uniqueid_for_keyname();
```

Comment:
the builder pattern feels really awkward here. `std::string work` has a different meaning between the steps, and each step has to check `failed` instead of returning errors directly. can we refactor KmipGetTheKey into static functions that return int and take their state as arguments?

Reply:
it's kind of the basic idea of the class;  work is an internal state;  it would be similar to object to the fact that kmip_client is basically std C, constantly allocating objects and buffers that you wouldn't in modern c++--I'm tempted, but it seems better to ask "does it work?" and then valgrind it